docs/prod-host-setup.md

```diff
@@ -328,3 +328,78 @@ wget -O - https://meltdown.ovh | bash
 
 [1]: https://elixir.free-electrons.com/linux/v4.14.203/source/virt/kvm/arm/hyp/timer-sr.c#L63
 [2]: https://lists.cs.columbia.edu/pipermail/kvmarm/2017-January/023323.html
+
+### Linux 6.1 boot time regressions
+
+Linux 6.1 introduced some regressions in the time it takes to boot a VM, for the
+x86_64 architecture. They can be mitigated depending on the CPU and the version
+of cgroups in use.
+
+#### Explanation
+
+The regression happens in the `KVM_CREATE_VM` ioctl and there are two factors
+that cause the issue:
+
+1. In the implementation of the mitigation for the iTLB multihit vulnerability,
+   KVM creates a worker thread called `kvm-nx-lpage-recovery`. This thread is
+   responsible for recovering huge pages split when the mitigation kicks-in. In
+   the process of creating this thread, KVM calls `cgroup_attach_task_all()` to
+   move it to the same cgroup used by the hypervisor thread
+1. In kernel v4.4, upstream converted a cgroup per process read-write semaphore
+   into a per-cpu read-write semaphore to allow to perform operations across
+   multiple processes
+   ([commit](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?&id=1ed1328792ff46e4bb86a3d7f7be2971f4549f6c)).
+   It was found that this conversion introduced high latency for write paths,
+   which mainly includes moving tasks between cgroups. This was fixed in kernel
+   v4.9 by
+   [commit](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?&id=3942a9bd7b5842a924e99ee6ec1350b8006c94ec)
+   which chose to favor writers over readers since moving tasks between cgroups
+   is a common operation for Android. However, In kernel 6.0, upstream decided
+   to revert back again and favor readers over writers re-introducing the
+   original behavior of the rw semaphore
+   ([commit](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?&id=6a010a49b63ac8465851a79185d8deff966f8e1a)).
+   At the same time, this commit provided an option called favordynmods to favor
+   writers over readers.
+1. Since the `kvm-nx-lpage-recovery` thread creation and its cgroup change is done
+   in the `KVM_CREATE_VM` call, the high latency we observe in 6.1 is due to the
+   upstream decision to favor readers over writers for this per-cpu rw
+   semaphore. While the 4.14 and 5.10 kernels favor writers over readers.
+
+#### If the host is vulnerable to iTLB multihit
+
+To check if the host is vulnerable, look at the value of
+`/sys/devices/system/cpu/vulnerabilities/itlb_multihit`. If it does not say `Not
+affected`, the host is vulnerable.
```

Comment:
I think right now it says `Affected`, but I wanted to future proof it in case there is a software mitigation. Yeah not satisfied with the double negative either. Let me have another go at it.

Reply:
Moved this before and restructured a bit the sections and reworded. Let me know if this works better.